File: cli/src/commands/notifications.ts
import chalk from 'chalk'
import { requireApiKey } from '../config.js'
import { api } from '../api.js'
import { accent, accentBold, dim } from '../theme.js'

interface Notification {
  id: string
  type: string
  referenceId: string | null
  actorDisplay: string | null
  read: boolean
  createdAt: string
}

const TYPE_LABELS: Record<string, string> = {
  reply: 'replied to your post',
  mention: 'mentioned you',
  follow: 'followed you',
  like: 'upvoted your post',
  dm: 'sent you a message',
  quote: 'quoted your post',
  repost: 'reposted your post',
}

export async function notificationsCommand(options: { read?: boolean }): Promise<void> {
  const key = requireApiKey()

  if (options.read) {
    await api('/notifications', key, { method: 'POST' })
    console.log(dim('All notifications marked as read.'))
    return
  }

  const res = await api<{ notifications: Notification[]; unreadCount: number }>(
    '/notifications',
    key
  )
  const { notifications, unreadCount } = res.data!

  console.log()
  if (unreadCount > 0) {
    console.log(accentBold(`${unreadCount} unread`))
  }

  if (notifications.length === 0) {
    console.log(dim('No notifications.'))
    console.log()
    return
  }

  console.log()
  for (const n of notifications) {
    const actor = n.actorDisplay ? accent(n.actorDisplay) : dim('Someone')
    const action = TYPE_LABELS[n.type] ?? n.type
    const unread = n.read ? '' : chalk.yellow(' *')
    const ref = n.referenceId ? dim(` (${n.referenceId.slice(0, 8)}...)`) : ''
    const time = new Date(n.createdAt).toLocaleDateString('en-US', {
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    })

    console.log(`  ${actor} ${action}${ref}${unread}  ${dim(time)}`)
  }
  console.log()
  console.log(dim('Use --read to mark all as read'))
  console.log()
}

File: cli-rust/src/commands/notifications.rs
use anyhow::Result;
use serde::Deserialize;

use crate::api::ArkoraApi;
use crate::config;
use crate::theme;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct NotifData {
    notifications: Vec<Notification>,
    unread_count: i64,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Notification {
    #[serde(rename = "type")]
    notif_type: String,
    reference_id: Option<String>,
    actor_display: Option<String>,
    read: bool,
    created_at: String,
}

fn type_label(t: &str) -> &str {
    match t {
        "reply" => "replied to your post",
        "mention" => "mentioned you",
        "follow" => "followed you",
        "like" => "upvoted your post",
        "dm" => "sent you a message",
        "quote" => "quoted your post",
        "repost" => "reposted your post",
        _ => t,
    }
}

pub async fn run(mark_read: bool, color: (u8, u8, u8)) -> Result<()> {
    let cfg = config::load();
    let key = config::require_key(&cfg);
    let api = ArkoraApi::new(&config::api_url(&cfg), &key);

    if mark_read {
        api.post::<serde_json::Value>("/notifications", &serde_json::json!({}))
            .await?;
        println!("{}", theme::dim("All notifications marked as read."));
        return Ok(());
    }

    let resp = api.get::<NotifData>("/notifications").await?;
    let data = resp.data.ok_or_else(|| anyhow::anyhow!("No data in response"))?;

    println!();
    if data.unread_count > 0 {
        println!(
            "{}",
            theme::accent_bold(&format!("{} unread", data.unread_count), color)
        );
    }

    if data.notifications.is_empty() {
        println!("{}", theme::dim("No notifications."));
        println!();
        return Ok(());
    }

    println!();
    for n in &data.notifications {
        let actor = n
            .actor_display
            .as_deref()
            .map(|a| theme::accent(a, color))
            .unwrap_or_else(|| theme::dim("Someone"));
        let action = type_label(&n.notif_type);
        let unread = if n.read { "" } else { " *" };
        let reference = n
            .reference_id
            .as_deref()
            .map(|r| format!(" ({}...)", &r[..r.len().min(8)]))
            .unwrap_or_default();
        let time = n.created_at.get(..10).unwrap_or(&n.created_at);

        println!(
            "  {actor} {action}{ref_dim}{unread_mark}  {time_dim}",
            ref_dim = theme::dim(&reference),
            unread_mark = if unread.is_empty() {
                String::new()
            } else {
                " *".to_string()
            },
            time_dim = theme::dim(time)
        );
    }
    println!();
    println!("{}", theme::dim("Use --read to mark all as read"));
    println!();

    Ok(())
}
